CustomCollections/CustomCollections/CustomHashTable.cs

```diff
@@ -0,0 +1,220 @@
+            }
+            if (!HasAvailableMemory()) IncreaseSize();
+            if (ContainsKey(item.Key))
+            {
+                throw new ArgumentException("key");
+            }
+            var position = GetPosition(item.Key);
+            _hashTable[position] = item;
+            Count++;
+        }
+
+        public void Add(TKey key, TValue value)
+        {
+            Add(new KeyValuePair<TKey, TValue>(key, value));
+        }
+
+        public bool Remove(KeyValuePair<TKey, TValue> item)
+        {
+            return Remove(item.Key);
+        }
+
+        public bool Remove(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("key");
+            }
+            var position = GetPosition(key);
+            if (_hashTable[position].Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(_hashTable[position], new KeyValuePair<TKey, TValue>()))
+            {
+                _hashTable[position] = new KeyValuePair<TKey, TValue>();
+                Count--;
+                return true;
+            }
+            return false;
+        }
+
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            foreach (var item in _hashTable)
+            {
+                if (!item.Equals(null) || !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    yield return item;
+                }
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private int GetPosition(TKey key)
+        {
+            var hash = key.GetHashCode();
+            var capacity = Capacity;
+            if (hash >= Capacity) capacity = hash + 1;
+            var pos = Math.Abs(hash % capacity);
+            return pos;
+        }
+
+        private ICollection<TKey> GetKeys()
+        {
+            List<TKey> keys = new List<TKey>();
+            foreach (var item in _hashTable)
+            {
+                if (item.Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    keys.Add(item.Key);
+                }
+            }
+            return keys;
+        }
+
+        private ICollection<TValue> GetValues()
+        {
+            List<TValue> values = new List<TValue>();
+            foreach (var item in _hashTable)
+            {
+                if (item.Value != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    values.Add(item.Value);
+                }
+            }
+            return values;
+        }
+
+        private bool HasAvailableMemory()
+        {
+            return Count + 1 < Capacity;
```

Comment:
What is the reason for +1 here? Generally hashtables have way bigger internal array size then capacity to reduce impact of collisions.